File: typescript/exercises/code-smells/large-class/src/UserManager.ts
/**
 * UserManager handles all user-related operations
 *
 * This class demonstrates the "Large Class" code smell by having too many responsibilities:
 * - User management (CRUD operations)
 * - Authentication
 * - Authorization
 * - Email sending
 * - Activity logging
 */

interface User {
  id: string;
  username: string;
  email: string;
  password: string;
  roles: string[];
  createdAt: string;
  updatedAt: string;
  lastLogin: string | null;
  loginCount: number;
  status: 'active' | 'inactive' | 'suspended';
  emailVerified: boolean;
  profile: {
    firstName: string;
    lastName: string;
    avatar: string;
    bio: string;
  };
}

interface Session {
  token: string;
  userId: string;
  createdAt: string;
  expiresAt: string;
  ipAddress: string;
  userAgent: string;
}

interface Role {
  name: string;
  permissions: string[];
}

interface ActivityLog {
  userId: string;
  action: string;
  description: string;
  timestamp: string;
  ipAddress: string;
}

interface EmailItem {
  to: string;
  subject: string;
  message: string;
  queuedAt: string;
}

export class UserManager {
  private users: Map<string, User> = new Map();
  private sessions: Map<string, Session> = new Map();
  private roles: Map<string, Role> = new Map();
  private permissions: Map<string, string> = new Map();
  private activityLog: ActivityLog[] = [];
  private emailQueue: EmailItem[] = [];
  private emailEnabled: boolean = true;
  private loggingEnabled: boolean = true;

  constructor() {
    this.initializeDefaultRoles();
    this.initializeDefaultPermissions();
  }

  // ==== USER MANAGEMENT ====

  /**
   * Creates a new user with validation and email notification
   */
  createUser(
    username: string,
    email: string,
    password: string,
    roles: string[] = ['user']
  ): User {
    // Validate username
    if (!username || username.length < 3) {
      throw new Error('Username must be at least 3 characters long');
    }

    if (this.getUserByUsername(username) !== null) {
      throw new Error('Username already exists');
    }

    // Validate email
    const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
    if (!emailRegex.test(email)) {
      throw new Error('Invalid email address');
    }

    if (this.getUserByEmail(email) !== null) {
      throw new Error('Email already registered');
    }

    // Validate password
    if (password.length < 8) {
      throw new Error('Password must be at least 8 characters long');
    }

    if (
      !/[A-Z]/.test(password) ||
      !/[a-z]/.test(password) ||
      !/[0-9]/.test(password)
    ) {
      throw new Error(
        'Password must contain uppercase, lowercase, and numbers'
      );
    }

    // Validate roles
    for (const role of roles) {
      if (!this.roles.has(role)) {
        throw new Error(`Invalid role: ${role}`);
      }
    }

    // Hash password (simplified for demo)
    const hashedPassword = this.hashPassword(password);

    // Create user
    const user: User = {
      id: this.generateId(),
      username,
      email,
      password: hashedPassword,
      roles,
      createdAt: new Date().toISOString(),
      updatedAt: new Date().toISOString(),
      lastLogin: null,
      loginCount: 0,
      status: 'active',
      emailVerified: false,
      profile: {
        firstName: '',
        lastName: '',
        avatar: '',
        bio: '',
      },
    };

    this.users.set(user.id, user);

    // Log activity
    this.logActivity(user.id, 'user_created', `User ${username} created`);

    // Send welcome email
    this.sendWelcomeEmail(email, username);

    return user;
  }

  /**
   * Updates user information with validation
   */
  updateUser(userId: string, data: Partial<User>): User {
    const user = this.getUserById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    // Validate and update username
    if (data.username !== undefined) {
      if (!data.username || data.username.length < 3) {
        throw new Error('Username must be at least 3 characters long');
      }

      const existingUser = this.getUserByUsername(data.username);
      if (existingUser && existingUser.id !== userId) {
        throw new Error('Username already exists');
      }

      user.username = data.username;
    }

    // Validate and update email
    if (data.email !== undefined) {
      const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
      if (!emailRegex.test(data.email)) {
        throw new Error('Invalid email address');
      }

      const existingUser = this.getUserByEmail(data.email);
      if (existingUser && existingUser.id !== userId) {
        throw new Error('Email already registered');
      }

      const oldEmail = user.email;
      user.email = data.email;
      user.emailVerified = false;

      // Send email change notification
      this.sendEmailChangeNotification(oldEmail, data.email);
    }

    // Update password if provided
    if (data.password !== undefined) {
      if (data.password.length < 8) {
        throw new Error('Password must be at least 8 characters long');
      }

      if (
        !/[A-Z]/.test(data.password) ||
        !/[a-z]/.test(data.password) ||
        !/[0-9]/.test(data.password)
      ) {
        throw new Error(
          'Password must contain uppercase, lowercase, and numbers'
        );
      }

      user.password = this.hashPassword(data.password);
    }

    // Update roles if provided
    if (data.roles !== undefined) {
      for (const role of data.roles) {
        if (!this.roles.has(role)) {
          throw new Error(`Invalid role: ${role}`);
        }
      }
      user.roles = data.roles;
    }

    // Update profile data
    if (data.profile !== undefined) {
      user.profile = { ...user.profile, ...data.profile };
    }

    // Update status
    if (data.status !== undefined) {
      if (!['active', 'inactive', 'suspended'].includes(data.status)) {
        throw new Error('Invalid status');
      }
      user.status = data.status;
    }

    user.updatedAt = new Date().toISOString();
    this.users.set(userId, user);

    // Log activity
    this.logActivity(userId, 'user_updated', `User ${user.username} updated`);

    return user;
  }

  /**
   * Deletes a user and cleans up associated data
   */
  deleteUser(userId: string): boolean {
    const user = this.getUserById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    // Log activity before deletion
    this.logActivity(userId, 'user_deleted', `User ${user.username} deleted`);

    // Send goodbye email
    this.sendGoodbyeEmail(user.email, user.username);

    // Clean up sessions
    this.cleanupUserSessions(userId);

    // Remove user
    this.users.delete(userId);

    return true;
  }

  // ==== AUTHENTICATION ====

  /**
   * Authenticates user and creates session
   */
  login(username: string, password: string): Session {
    const user = this.getUserByUsername(username);
    if (!user) {
      throw new Error('Invalid username or password');
    }

    if (user.status !== 'active') {
      throw new Error('Account is not active');
    }

    if (!this.verifyPassword(password, user.password)) {
      // Log failed login attempt
      this.logActivity(
        user.id,
        'login_failed',
        `Failed login attempt for ${username}`
      );
      throw new Error('Invalid username or password');
    }

    // Generate session token
    const sessionToken = this.generateSessionToken();
    const session: Session = {
      token: sessionToken,
      userId: user.id,
      createdAt: new Date().toISOString(),
      expiresAt: new Date(Date.now() + 24 * 60 * 60 * 1000).toISOString(), // 24 hours
      ipAddress: 'unknown', // Would be from request in real app
      userAgent: 'unknown', // Would be from request in real app
    };

    this.sessions.set(sessionToken, session);

    // Update user login info
    user.lastLogin = new Date().toISOString();
    user.loginCount++;
    this.users.set(user.id, user);

    // Log successful login
    this.logActivity(user.id, 'login_success', `User ${username} logged in`);

    // Send login notification email
    this.sendLoginNotificationEmail(
      user.email,
      user.username,
      session.ipAddress
    );

    return session;
  }

  /**
   * Validates session token and returns user data
   */
  validateSession(token: string): User | null {
    const session = this.sessions.get(token);
    if (!session) {
      return null;
    }

    // Check if session is expired
    if (new Date(session.expiresAt) < new Date()) {
      this.sessions.delete(token);
      return null;
    }

    const user = this.getUserById(session.userId);
    if (!user || user.status !== 'active') {
      this.sessions.delete(token);
      return null;
    }

    return user;
  }

  /**
   * Logs out user and destroys session
   */
  logout(token: string): boolean {
    const session = this.sessions.get(token);
    if (session) {
      const user = this.getUserById(session.userId);

      if (user) {
        this.logActivity(user.id, 'logout', `User ${user.username} logged out`);
      }

      this.sessions.delete(token);
      return true;
    }

    return false;
  }

  // ==== AUTHORIZATION ====

  /**
   * Checks if user has specific permission
   */
  hasPermission(userId: string, permission: string): boolean {
    const user = this.getUserById(userId);
    if (!user || user.status !== 'active') {
      return false;
    }

    for (const roleName of user.roles) {
      const role = this.roles.get(roleName);
      if (role && role.permissions.includes(permission)) {
        return true;
      }
    }

    return false;
  }

  /**
   * Checks if user has specific role
   */
  hasRole(userId: string, roleName: string): boolean {
    const user = this.getUserById(userId);
    if (!user) {
      return false;
    }

    return user.roles.includes(roleName);
  }

  /**
   * Assigns role to user
   */
  assignRole(userId: string, roleName: string): boolean {
    if (!this.roles.has(roleName)) {
      throw new Error(`Invalid role: ${roleName}`);
    }

    const user = this.getUserById(userId);
    if (!user) {
      throw new Error('User not found');
    }

    if (!user.roles.includes(roleName)) {
      user.roles.push(roleName);
      this.users.set(userId, user);
      this.logActivity(
        userId,
        'role_assigned',
        `Role ${roleName} assigned to user ${user.username}`
      );
    }

    return true;
  }

  // ==== EMAIL OPERATIONS ====

  /**
   * Sends welcome email to new user
   */
  private sendWelcomeEmail(email: string, username: string): void {
    if (!this.emailEnabled) {
      return;
    }

    const subject = 'Welcome to our platform!';
    const message = `Hello ${username},\n\nWelcome to our platform! Your account has been created successfully.\n\nBest regards,\nThe Team`;

    this.queueEmail(email, subject, message);
  }

  /**
   * Sends email change notification
   */
  private sendEmailChangeNotification(
    oldEmail: string,
    newEmail: string
  ): void {
    if (!this.emailEnabled) {
      return;
    }

    const subject = 'Email address changed';
    const message = `Your email address has been changed from ${oldEmail} to ${newEmail}.\n\nIf you didn't make this change, please contact support immediately.`;

    this.queueEmail(oldEmail, subject, message);
    this.queueEmail(newEmail, subject, message);
  }

  /**
   * Sends login notification email
   */
  private sendLoginNotificationEmail(
    email: string,
    username: string,
    ipAddress: string
  ): void {
    if (!this.emailEnabled) {
      return;
    }

    const subject = 'New login detected';
    const message = `Hello ${username},\n\nA new login was detected from IP address: ${ipAddress}\n\nIf this wasn't you, please change your password immediately.`;

    this.queueEmail(email, subject, message);
  }

  /**
   * Sends goodbye email when user is deleted
   */
  private sendGoodbyeEmail(email: string, username: string): void {
    if (!this.emailEnabled) {
      return;
    }

    const subject = 'Account deleted';
    const message = `Hello ${username},\n\nYour account has been deleted. We're sorry to see you go!\n\nBest regards,\nThe Team`;

    this.queueEmail(email, subject, message);
  }

  /**
   * Queues email for sending
   */
  private queueEmail(to: string, subject: string, message: string): void {
    this.emailQueue.push({
      to,
      subject,
      message,
      queuedAt: new Date().toISOString(),
    });
  }

  // ==== ACTIVITY LOGGING ====

  /**
   * Logs user activity
   */
  private logActivity(
    userId: string,
    action: string,
    description: string
  ): void {
    if (!this.loggingEnabled) {
      return;
    }

    this.activityLog.push({
      userId,
      action,
      description,
      timestamp: new Date().toISOString(),
      ipAddress: 'unknown', // Would be from request in real app
    });
  }

  /**
   * Gets activity log for specific user
   */
  getUserActivityLog(userId: string): ActivityLog[] {
    return this.activityLog.filter(log => log.userId === userId);
  }

  // ==== HELPER METHODS ====

  getUserById(id: string): User | null {
    return this.users.get(id) || null;
  }

  getUserByUsername(username: string): User | null {
    for (const user of this.users.values()) {
      if (user.username === username) {
        return user;
      }
    }
    return null;
  }

  getUserByEmail(email: string): User | null {
    for (const user of this.users.values()) {
      if (user.email === email) {
        return user;
      }
    }
    return null;
  }

  getAllUsers(): User[] {
    return Array.from(this.users.values());
  }

  getUserCount(): number {
    return this.users.size;
  }

  private cleanupUserSessions(userId: string): void {
    for (const [token, session] of this.sessions.entries()) {
      if (session.userId === userId) {
        this.sessions.delete(token);
      }
    }
  }

  private initializeDefaultRoles(): void {
    this.roles.set('admin', {
      name: 'Administrator',
      permissions: [
        'user_create',
        'user_read',
        'user_update',
        'user_delete',
        'admin_panel',
      ],
    });

    this.roles.set('moderator', {
      name: 'Moderator',
      permissions: ['user_read', 'user_update', 'moderate_content'],
    });

    this.roles.set('user', {
      name: 'User',
      permissions: ['user_read', 'profile_update'],
    });
  }

  private initializeDefaultPermissions(): void {
    this.permissions.set('user_create', 'Create new users');
    this.permissions.set('user_read', 'View user information');
    this.permissions.set('user_update', 'Update user information');
    this.permissions.set('user_delete', 'Delete users');
    this.permissions.set('admin_panel', 'Access admin panel');
    this.permissions.set('moderate_content', 'Moderate user content');
    this.permissions.set('profile_update', 'Update own profile');
  }

  // Utility methods
  private generateId(): string {
    return Math.random().toString(36).substr(2, 9);
  }

  private generateSessionToken(): string {
    return Math.random().toString(36).substr(2, 32);
  }

  private hashPassword(password: string): string {
    // Simplified hashing for demo purposes
    return `hashed_${password}`;
  }

  private verifyPassword(password: string, hashedPassword: string): boolean {
    return hashedPassword === `hashed_${password}`;
  }

  // Configuration methods
  setEmailEnabled(enabled: boolean): void {
    this.emailEnabled = enabled;
  }

  setLoggingEnabled(enabled: boolean): void {
    this.loggingEnabled = enabled;
  }

  getEmailQueue(): EmailItem[] {
    return this.emailQueue;
  }

  getActivityLog(): ActivityLog[] {
    return this.activityLog;
  }
}

File: python/exercises/code-smells/large-class/src/user_manager.py
"""
UserManager handles all user-related operations

This class demonstrates the "Large Class" code smell by having too many responsibilities:
- User management (CRUD operations)
- Authentication
- Authorization
- Email sending
- Activity logging
"""

import re
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timedelta
from typing import Any


@dataclass
class User:
    id: str
    username: str
    email: str
    password: str
    roles: list[str]
    created_at: str
    updated_at: str
    last_login: str | None
    login_count: int
    status: str
    email_verified: bool
    profile: dict[str, str] = field(
        default_factory=lambda: {
            "first_name": "",
            "last_name": "",
            "avatar": "",
            "bio": "",
        }
    )


@dataclass
class Session:
    token: str
    user_id: str
    created_at: str
    expires_at: str
    ip_address: str
    user_agent: str


@dataclass
class Role:
    name: str
    permissions: list[str]


@dataclass
class ActivityLog:
    user_id: str
    action: str
    description: str
    timestamp: str
    ip_address: str


@dataclass
class EmailItem:
    to: str
    subject: str
    message: str
    queued_at: str


class UserManager:
    def __init__(self):
        self.users: dict[str, User] = {}
        self.sessions: dict[str, Session] = {}
        self.roles: dict[str, Role] = {}
        self.permissions: dict[str, str] = {}
        self.activity_log: list[ActivityLog] = []
        self.email_queue: list[EmailItem] = []
        self.email_enabled: bool = True
        self.logging_enabled: bool = True

        self._initialize_default_roles()
        self._initialize_default_permissions()

    # ==== USER MANAGEMENT ====

    def create_user(
        self, username: str, email: str, password: str, roles: list[str] = None
    ) -> User:
        """Creates a new user with validation and email notification"""
        if roles is None:
            roles = ["user"]

        # Validate username
        if not username or len(username) < 3:
            raise ValueError("Username must be at least 3 characters long")

        if self.get_user_by_username(username) is not None:
            raise ValueError("Username already exists")

        # Validate email
        email_pattern = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"
        if not re.match(email_pattern, email):
            raise ValueError("Invalid email address")

        if self.get_user_by_email(email) is not None:
            raise ValueError("Email already registered")

        # Validate password
        if len(password) < 8:
            raise ValueError("Password must be at least 8 characters long")

        if not (
            re.search(r"[A-Z]", password)
            and re.search(r"[a-z]", password)
            and re.search(r"[0-9]", password)
        ):
            raise ValueError("Password must contain uppercase, lowercase, and numbers")

        # Validate roles
        for role in roles:
            if role not in self.roles:
                raise ValueError(f"Invalid role: {role}")

        # Hash password
        hashed_password = self._hash_password(password)

        # Create user
        user = User(
            id=str(uuid.uuid4()),
            username=username,
            email=email,
            password=hashed_password,
            roles=roles,
            created_at=datetime.now().isoformat(),
            updated_at=datetime.now().isoformat(),
            last_login=None,
            login_count=0,
            status="active",
            email_verified=False,
        )

        self.users[user.id] = user

        # Log activity
        self._log_activity(user.id, "user_created", f"User {username} created")

        # Send welcome email
        self._send_welcome_email(email, username)

        return user

    def update_user(self, user_id: str, data: dict[str, Any]) -> User:
        """Updates user information with validation"""
        user = self.get_user_by_id(user_id)
        if user is None:
            raise ValueError("User not found")

        # Validate and update username
        if "username" in data:
            if not data["username"] or len(data["username"]) < 3:
                raise ValueError("Username must be at least 3 characters long")

            existing_user = self.get_user_by_username(data["username"])
            if existing_user is not None and existing_user.id != user_id:
                raise ValueError("Username already exists")

            user.username = data["username"]

        # Validate and update email
        if "email" in data:
            email_pattern = r"^[^\s@]+@[^\s@]+\.[^\s@]+$"
            if not re.match(email_pattern, data["email"]):
                raise ValueError("Invalid email address")

            existing_user = self.get_user_by_email(data["email"])
            if existing_user is not None and existing_user.id != user_id:
                raise ValueError("Email already registered")

            old_email = user.email
            user.email = data["email"]
            user.email_verified = False

            # Send email change notification
            self._send_email_change_notification(old_email, data["email"])

        # Update password if provided
        if "password" in data:
            if len(data["password"]) < 8:
                raise ValueError("Password must be at least 8 characters long")

            if not (
                re.search(r"[A-Z]", data["password"])
                and re.search(r"[a-z]", data["password"])
                and re.search(r"[0-9]", data["password"])
            ):
                raise ValueError(
                    "Password must contain uppercase, lowercase, and numbers"
                )

            user.password = self._hash_password(data["password"])

        # Update roles if provided
        if "roles" in data:
            for role in data["roles"]:
                if role not in self.roles:
                    raise ValueError(f"Invalid role: {role}")
            user.roles = data["roles"]

        # Update profile data
        if "profile" in data:
            user.profile.update(data["profile"])

        # Update status
        if "status" in data:
            if data["status"] not in ["active", "inactive", "suspended"]:
                raise ValueError("Invalid status")
            user.status = data["status"]

        user.updated_at = datetime.now().isoformat()
        self.users[user_id] = user

        # Log activity
        self._log_activity(user_id, "user_updated", f"User {user.username} updated")

        return user

    def delete_user(self, user_id: str) -> bool:
        """Deletes a user and cleans up associated data"""
        user = self.get_user_by_id(user_id)
        if user is None:
            raise ValueError("User not found")

        # Log activity before deletion
        self._log_activity(user_id, "user_deleted", f"User {user.username} deleted")

        # Send goodbye email
        self._send_goodbye_email(user.email, user.username)

        # Clean up sessions
        self._cleanup_user_sessions(user_id)

        # Remove user
        del self.users[user_id]

        return True

    # ==== AUTHENTICATION ====

    def login(self, username: str, password: str) -> Session:
        """Authenticates user and creates session"""
        user = self.get_user_by_username(username)
        if user is None:
            raise ValueError("Invalid username or password")

        if user.status != "active":
            raise RuntimeError("Account is not active")

        if not self._verify_password(password, user.password):
            # Log failed login attempt
            self._log_activity(
                user.id, "login_failed", f"Failed login attempt for {username}"
            )
            raise ValueError("Invalid username or password")

        # Generate session token
        session_token = self._generate_session_token()
        session = Session(
            token=session_token,
            user_id=user.id,
            created_at=datetime.now().isoformat(),
            expires_at=(datetime.now() + timedelta(hours=24)).isoformat(),
            ip_address="unknown",  # Would be from request in real app
            user_agent="unknown",  # Would be from request in real app
        )

        self.sessions[session_token] = session

        # Update user login info
        user.last_login = datetime.now().isoformat()
        user.login_count += 1
        self.users[user.id] = user

        # Log successful login
        self._log_activity(user.id, "login_success", f"User {username} logged in")

        # Send login notification email
        self._send_login_notification_email(
            user.email, user.username, session.ip_address
        )

        return session

    def validate_session(self, token: str) -> User | None:
        """Validates session token and returns user data"""
        if token not in self.sessions:
            return None

        session = self.sessions[token]

        # Check if session is expired
        if datetime.fromisoformat(session.expires_at) < datetime.now():
            del self.sessions[token]
            return None

        user = self.get_user_by_id(session.user_id)
        if user is None or user.status != "active":
            del self.sessions[token]
            return None

        return user

    def logout(self, token: str) -> bool:
        """Logs out user and destroys session"""
        if token in self.sessions:
            session = self.sessions[token]
            user = self.get_user_by_id(session.user_id)

            if user is not None:
                self._log_activity(
                    user.id, "logout", f"User {user.username} logged out"
                )

            del self.sessions[token]
            return True

        return False

    # ==== AUTHORIZATION ====

    def has_permission(self, user_id: str, permission: str) -> bool:
        """Checks if user has specific permission"""
        user = self.get_user_by_id(user_id)
        if user is None or user.status != "active":
            return False

        for role_name in user.roles:
            if role_name in self.roles:
                role = self.roles[role_name]
                if permission in role.permissions:
                    return True

        return False

    def has_role(self, user_id: str, role_name: str) -> bool:
        """Checks if user has specific role"""
        user = self.get_user_by_id(user_id)
        if user is None:
            return False

        return role_name in user.roles

    def assign_role(self, user_id: str, role_name: str) -> bool:
        """Assigns role to user"""
        if role_name not in self.roles:
            raise ValueError(f"Invalid role: {role_name}")

        user = self.get_user_by_id(user_id)
        if user is None:
            raise ValueError("User not found")

        if role_name not in user.roles:
            user.roles.append(role_name)
            self.users[user_id] = user
            self._log_activity(
                user_id,
                "role_assigned",
                f"Role {role_name} assigned to user {user.username}",
            )

        return True

    # ==== EMAIL OPERATIONS ====

    def _send_welcome_email(self, email: str, username: str) -> None:
        """Sends welcome email to new user"""
        if not self.email_enabled:
            return

        subject = "Welcome to our platform!"
        message = f"Hello {username},\n\nWelcome to our platform! Your account has been created successfully.\n\nBest regards,\nThe Team"

        self._queue_email(email, subject, message)

    def _send_email_change_notification(self, old_email: str, new_email: str) -> None:
        """Sends email change notification"""
        if not self.email_enabled:
            return

        subject = "Email address changed"
        message = f"Your email address has been changed from {old_email} to {new_email}.\n\nIf you didn't make this change, please contact support immediately."

        self._queue_email(old_email, subject, message)
        self._queue_email(new_email, subject, message)

    def _send_login_notification_email(
        self, email: str, username: str, ip_address: str
    ) -> None:
        """Sends login notification email"""
        if not self.email_enabled:
            return

        subject = "New login detected"
        message = f"Hello {username},\n\nA new login was detected from IP address: {ip_address}\n\nIf this wasn't you, please change your password immediately."

        self._queue_email(email, subject, message)

    def _send_goodbye_email(self, email: str, username: str) -> None:
        """Sends goodbye email when user is deleted"""
        if not self.email_enabled:
            return

        subject = "Account deleted"
        message = f"Hello {username},\n\nYour account has been deleted. We're sorry to see you go!\n\nBest regards,\nThe Team"

        self._queue_email(email, subject, message)

    def _queue_email(self, to: str, subject: str, message: str) -> None:
        """Queues email for sending"""
        self.email_queue.append(
            EmailItem(
                to=to,
                subject=subject,
                message=message,
                queued_at=datetime.now().isoformat(),
            )
        )

    # ==== ACTIVITY LOGGING ====

    def _log_activity(self, user_id: str, action: str, description: str) -> None:
        """Logs user activity"""
        if not self.logging_enabled:
            return

        self.activity_log.append(
            ActivityLog(
                user_id=user_id,
                action=action,
                description=description,
                timestamp=datetime.now().isoformat(),
                ip_address="unknown",  # Would be from request in real app
            )
        )

    def get_user_activity_log(self, user_id: str) -> list[ActivityLog]:
        """Gets activity log for specific user"""
        return [log for log in self.activity_log if log.user_id == user_id]

    # ==== HELPER METHODS ====

    def get_user_by_id(self, user_id: str) -> User | None:
        """Gets user by ID"""
        return self.users.get(user_id)

    def get_user_by_username(self, username: str) -> User | None:
        """Gets user by username"""
        for user in self.users.values():
            if user.username == username:
                return user
        return None

    def get_user_by_email(self, email: str) -> User | None:
        """Gets user by email"""
        for user in self.users.values():
            if user.email == email:
                return user
        return None

    def get_all_users(self) -> list[User]:
        """Gets all users"""
        return list(self.users.values())

    def get_user_count(self) -> int:
        """Gets user count"""
        return len(self.users)

    def _cleanup_user_sessions(self, user_id: str) -> None:
        """Cleans up sessions for a user"""
        tokens_to_remove = []
        for token, session in self.sessions.items():
            if session.user_id == user_id:
                tokens_to_remove.append(token)

        for token in tokens_to_remove:
            del self.sessions[token]

    def _initialize_default_roles(self) -> None:
        """Initializes default roles"""
        self.roles = {
            "admin": Role(
                name="Administrator",
                permissions=[
                    "user_create",
                    "user_read",
                    "user_update",
                    "user_delete",
                    "admin_panel",
                ],
            ),
            "moderator": Role(
                name="Moderator",
                permissions=["user_read", "user_update", "moderate_content"],
            ),
            "user": Role(name="User", permissions=["user_read", "profile_update"]),
        }

    def _initialize_default_permissions(self) -> None:
        """Initializes default permissions"""
        self.permissions = {
            "user_create": "Create new users",
            "user_read": "View user information",
            "user_update": "Update user information",
            "user_delete": "Delete users",
            "admin_panel": "Access admin panel",
            "moderate_content": "Moderate user content",
            "profile_update": "Update own profile",
        }

    # Utility methods
    def _generate_id(self) -> str:
        """Generates a unique ID"""
        return str(uuid.uuid4())

    def _generate_session_token(self) -> str:
        """Generates a session token"""
        return str(uuid.uuid4()).replace("-", "")

    def _hash_password(self, password: str) -> str:
        """Hashes a password (simplified for demo)"""
        return f"hashed_{password}"

    def _verify_password(self, password: str, hashed_password: str) -> bool:
        """Verifies a password against its hash"""
        return hashed_password == f"hashed_{password}"

    # Configuration methods
    def set_email_enabled(self, enabled: bool) -> None:
        """Sets email enabled state"""
        self.email_enabled = enabled

    def set_logging_enabled(self, enabled: bool) -> None:
        """Sets logging enabled state"""
        self.logging_enabled = enabled

    def get_email_queue(self) -> list[EmailItem]:
        """Gets email queue"""
        return self.email_queue

    def get_activity_log(self) -> list[ActivityLog]:
        """Gets activity log"""
        return self.activity_log
